lib/common/service-object.js

```diff
@@ -0,0 +1,303 @@
+ * @param {string} config.baseUrl - The base URL to make API requests to.
+ * @param {string} config.createMethod - The method which creates this object.
+ * @param {string} config.id - The identifier of the object. For example, the
+ *     name of a Storage bucket or Pub/Sub topic.
+ * @param {object=} config.methods - A map of each method name that should be
+ *     inherited.
+ * @param {object} config.parent - The parent service instance. For example, an
+ *     instance of Storage if the object is Bucket.
+ */
+function ServiceObject(config) {
+  var self = this;
+
+  this.metadata = {};
+
+  this.baseUrl = config.baseUrl;
+  this.parent = config.parent; // Parent class.
+  this.id = config.id; // Name or ID (e.g. dataset ID, bucket name, etc.)
+  this.createMethod = config.createMethod;
+
+  if (config.methods) {
+    var allMethodNames = Object.keys(ServiceObject.prototype);
+    allMethodNames
+      .filter(function(methodName) {
+        return (
+          // All ServiceObjects need `request`.
+          methodName !== 'request' &&
+
+          // The ServiceObject didn't redefine the method.
+          self[methodName] === ServiceObject.prototype[methodName] &&
+
+          // This method isn't wanted.
+          !config.methods[methodName]
+        );
+      })
+      .forEach(function(methodName) {
+        self[methodName] = undefined;
+      });
+  }
+}
+
+/**
+ * Create the object.
+ *
+ * @param {object=} options - Configuration object.
+ * @param {function=} callback - The callback function.
+ * @param {?error} callback.err - An error returned while making this request.
+ * @param {object} callback.instance - The instance.
+ * @param {object} callback.apiResponse - The full API response.
+ */
+ServiceObject.prototype.create = function(options, callback) {
+  var self = this;
+  var args = [this.id];
+
+  if (is.fn(options)) {
+    callback = options;
+  }
+
+  if (is.object(options)) {
+    args.push(options);
+  }
+
+  // Wrap the callback to return *this* instance of the object, not the newly-
+  // created one.
+  function onCreate(err, instance, apiResponse) {
+    if (err) {
+      callback(err, null, apiResponse);
+      return;
+    }
+
+    self.metadata = instance.metadata;
+
+    callback(null, self, apiResponse);
+  }
+
+  args.push(onCreate);
+
+  this.createMethod.apply(null, args);
+};
+
+/**
+ * Delete the object.
+ *
+ * @param {function=} callback - The callback function.
+ * @param {?error} callback.err - An error returned while making this request.
+ * @param {object} callback.apiResponse - The full API response.
+ */
+ServiceObject.prototype.delete = function(callback) {
+  var reqOpts = {
+    method: 'DELETE',
+    uri: ''
+  };
+
+  // The `request` method may have been overridden to hold any special behavior.
+  // Ensure we call the original `request` method.
+  ServiceObject.prototype.request.call(this, reqOpts, function(err, resp) {
+    callback(err, resp);
```

Comment:
Haha, you did, and it's because this all patches back to `request`, which gives a long request/response object with details of the headers, etc. Instead of documenting yet another argument to all of our methods' callbacks, if there's an error, it's attached to the error object as `response`.